Maple2.Server.Game/PacketHandlers/ItemExchangeScrollHandler.cs

```diff
@@ -0,0 +1,159 @@
+                        }
+                    }
+                }
+
+                if (remaining > 0) {
+                    session.Send(ItemExchangeScrollPacket.Error(ItemExchangeScrollError.s_itemslot_exchange_count_invalid));
+                    return;
+                }
+            }
+            
+            // Consume scrolls
+            int remainingScrolls = scrollMetadata.RecipeScroll.Amount * quantity;
+            foreach (Item scroll in scrolls) {
+                int consume = Math.Min(remainingScrolls, scroll.Amount);
+                if (!session.Item.Inventory.Consume(scroll.Uid, consume)) {
+                    Logger.Fatal("Failed to consume item {ItemUid}", scroll.Uid);
+                    throw new InvalidOperationException($"Fatal: Consuming item: {scroll.Uid}");
+                }
+            }
+
+            // Consume ingredients
+            foreach (Ingredient ingredient in scrollMetadata.RequiredItems) {
+                int remainingIngredients = ingredient.Amount * quantity;
+                if (ingredient.Tag != ItemTag.None) {
+                    foreach (Item material in materialsByTag[ingredient.Tag]) {
+                        int consume = Math.Min(remainingIngredients, material.Amount);
+                        if (!session.Item.Inventory.Consume(material.Uid, consume)) {
+                            Logger.Fatal("Failed to consume item {ItemUid}", material.Uid);
+                            throw new InvalidOperationException($"Fatal: Consuming item: {material.Uid}");
+                        }
+
+                        remainingIngredients -= consume;
+                        if (remainingIngredients <= 0) {
+                            break;
+                        }
+                    }
+                } else {
+                    foreach (Item material in materialsById[ingredient.ItemId]) {
+                        int consume = Math.Min(remainingIngredients, material.Amount);
+                        if (!session.Item.Inventory.Consume(material.Uid, consume)) {
+                            Logger.Fatal("Failed to consume item {ItemUid}", material.Uid);
+                            throw new InvalidOperationException($"Fatal: Consuming item: {material.Uid}");
+                        }
+
+                        remainingIngredients -= consume;
+                        if (remainingIngredients <= 0) {
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        
+        session.Currency.Meso -= scrollMetadata.RequiredMeso;
+
+        if (!ItemMetadata.TryGet(scrollMetadata.RewardItem.ItemId, out ItemMetadata? itemMetadata)) {
+            return;
+        }
+            
+        var rewardItem = new Item(itemMetadata, scrollMetadata.RewardItem.Rarity, scrollMetadata.RewardItem.Amount * quantity);
+        if (!session.Item.Inventory.Add(rewardItem, true)) {
+            // TODO: mail player
```

Comment:
there should be a return here and/or some error otherwise it will be treated as success

Reply:
That's intentional. It's still a success, being mailed to the player is the only difference.

Reply:
Since it's not currently being mailed I felt like a failure would make more sense but it's fine.